velox/type/DecimalUtil.h

```diff
@@ -271,6 +271,37 @@ class DecimalUtil {
     return overflow;
   }
 
+  /// This method is used to correct the sum result calculated using
+  /// addWithOverflow. In the addWithOverflow method, when summing two int128_t
+  /// with the same sign, it first converts them to uint128_t for calculation.
+  /// The returned sum represents the lower 127 bits of the uint128_t sum.
+  /// The overflow variable indicates whether an overflow occurred during the
+  /// calculation. If overflow = 1, it means there was an upward overflow, and
+  /// if overflow = -1, there was a downward overflow. However, since the sum
+  /// only retains the lower 127 bits, it may miss one calculation of
+  /// +(1 << 127) or -(1 << 127).
+  /// Therefore, we need to make the following adjustments:
+  /// 1. If overflow = 1 && sum < 0, it means that the calculation process
+  /// missed one calculation of +(1 << 127), it needs to be corrected to
+  /// +(1 << 127) + sum.
+  /// 2. If overflow = -1 && sum > 0, it means that the calculation process
+  /// missed one calculation of -(1 << 127), it needs to be corrected to
+  /// -(1 << 127) + sum.
+  inline static std::optional<int128_t> adjustSumForOverflow(
+      int128_t sum,
+      int64_t overflow) {
+    // Value is valid if the conditions below are true.
+    if ((overflow == 1 && sum < 0) || (overflow == -1 && sum > 0)) {
+      return static_cast<int128_t>(
+          DecimalUtil::kOverflowMultiplier * overflow + sum);
+    }
+    if (overflow != 0) {
```

Comment:
Add a comment to explain what's going on here.

Should we throw user error instead of returning std::nullopt?

Should we also call valueInRange before returning a value?

This would make it easier for the caller as they call this method and either get an error or a value that can be used as is.

Reply:
> Should we throw user error instead of returning std::nullopt?

The reason for returning `std::nullopt` here is that Spark does not want to receive exception but wants to obtain overflow information through `std::nullopt` and return a null value instead.
This is also one of the practical scenarios we are discussing in https://github.com/facebookincubator/velox/discussions/7358.

> Should we also call valueInRange before returning a value?

Yes, we can move the `valueInRange` function inside this method because it is only used for Decimal calculations.